Support testing storage event triggered ADF pipelines

thurstonchen opened the pull request:
Resolving [issue #9 Support testing storage event triggered ADF pipelines](https://github.com/devlace/pytest-adf/issues/9); 
for more details, please check the mentioned issue, many thanks! :-)

thurstonchen committed:
feat: support testing storage event triggered ADF pipeline

thurstonchen committed:
feat: update README, requirements & setup.py to reflect the new feature

thurstonchen committed:
doc: update README to address storage account related env. variables

thurstonchen committed:
docs: update README.md

thurstonchen committed:
refactor: drop newly added polling method by reusing the existed one

devlace commented on README.md:
```diff
@@ -37,6 +37,50 @@ For an example of how to use this in an overall Modern Data Warehouse solution a
 
 For additional usage information, see [caching pipeline runs](#Caching-pipeline-runs).
 
```
Add a sub header `###` called "Testing Storage Event Triggered Pipelines" just to make it clearer that this is a different pytest fixture.

devlace commented on README.md:
````diff
@@ -37,6 +37,50 @@ For an example of how to use this in an overall Modern Data Warehouse solution a
 
+For testing storage event triggered pipelines, please use another fixture `adf_storage_event_triggered_pipeline_run`.
+
+```python
````
Add bolded title called **"Storage Event: BlobCreated".**
Also, would be good to explicitly set the `blob_event_type="BlobCreated"` parameter even if it is optional just to make it clear. I think it is enough that you also mention this in the additional notes.

```python
this_run = adf_storage_event_triggered_pipeline_run(PIPELINE_NAME,
                                                        TRIGGER_NAME,
                                                        TRIGGER_FILE_PATH,
                                                        CONTAINER_NAME,
                                                        LOCAL_TRIGGER_FILE_PATH,
                                                        blob_event_type="BlobCreated") # Default value.
```

devlace commented on README.md:
````diff
@@ -37,6 +37,50 @@ For an example of how to use this in an overall Modern Data Warehouse solution a
+- `BlobCreated` is the default blob event type, for `BlobDeleted` event trigger pipeline tests, please refer to below sample instead, which just skips the local-trigger-file-path argument and assigns `BlobDeleted` value to *blob_event_type* explicitly, then this fixture will help to delete declared event trigger file from the target container to trigger the specified pipeline run.
+
+
+```python
````
Move this up sample code up right after the BlobCreated sample code, but put another bolded title: **"Storage Event: BlobDeleted"**

I think if they are side by side it would be easier to spot the difference :)

devlace commented on src/pytest_adf/pytest_adf.py:
```diff
@@ -4,12 +4,28 @@
-OPTIONAL_ARGS = ["AZ_SERVICE_PRINCIPAL_ID", "AZ_SERVICE_PRINCIPAL_SECRET"]
+OPTIONAL_ARGS = ["AZ_SERVICE_PRINCIPAL_ID", "AZ_SERVICE_PRINCIPAL_SECRET", "AZ_STORAGE_ACCOUNT_NAME", "AZ_STORAGE_ACCOUNT_KEY"]
+OPTIONAL_STORAGE_ACCOUNT_ARGS = ["AZ_STORAGE_ACCOUNT_NAME", "AZ_STORAGE_ACCOUNT_KEY"]
+FETCHING_TRIGGER_RUNS_MAX_TIME_IN_SECONDS = 180
```
Add some comments about these constants. I was initially confused on the difference between the two and had to go through the code to figure it out :). 

How about this:
```python
# Determines the ADF query time range to search for the corresponding trigger ADF pipeline starting from when the blob file is uploaded/deleted. 
# For example, if file is uploaded/deleted at 01:00, this will search for the triggered pipelines between 01:00 and 01:03.
FETCHING_TRIGGER_RUNS_TIME_RANGE_IN_MINUTES = 180

# Determines the maximum amount of time pytest-adf will look for the triggered pipeline run within the FETCHING_TRIGGER_RUNS_TIME_RANGE_IN_MINUTES
FETCHING_TRIGGER_RUNS_MAX_TIME_IN_SECONDS = 3
```

devlace commented on src/pytest_adf/pytest_adf.py:
```diff
@@ -161,3 +208,162 @@ def _poll_adf_until(adf_client, az_resource_group, adf_name, pipeline_run_id,
+            trigger_runs = trigger_runs_query_response.value
+            if (len(trigger_runs) > 0):
+                for trigger_run in trigger_runs:
+                    event_payload = trigger_run.properties.get("EventPayload")
```
Perhaps add another `LOG.INFO` to log the resulting pipelines runs found.

Also, would be good to add an `else` block then add `LOG.INFO` that no pipelines have been found.

devlace reviewed (changes requested):
Looking good! I love that you are using the eTags for determining the correct storage event triggered which pipeline 👍 

Just some minor changes mostly around docs and also, if you run flake8 linting, there are number of linting errors. In the devcontainer you can run `make lint`.

Note to myself, need to update the PR validation pipelines :)

![image](https://user-images.githubusercontent.com/5771063/129301205-e3a42837-94ed-4a8b-9c0b-bff6f5359245.png)

thurstonchen committed:
docs: update README according to review comments

thurstonchen committed:
style: fix linting issues, and add info logs

thurstonchen commented:
Hi Lace, 
Thanks for your comments! 
I've updated them one by one in latest commits.
- Forgot the linting checks previously... anyway, they've been corrected now, thanks for the reminder!
- Updated README as suggested, now the test samples and descriptions for both blob created & deleted event triggers are not mixed anymore.
- Added suggested comments to those two constant variables.
- Added suggested info logs.

devlace reviewed (approved):
Great the changes look good! I've just approved the PR :)
